gen/constant.go

```diff
@@ -45,73 +45,97 @@ func Constant(g Generator, c *compile.Constant) error {
 func ConstantValue(g Generator, c compile.ConstantValue, t compile.TypeSpec) (string, error) {
 	switch v := c.(type) {
 	case compile.ConstantBool:
-		if v {
-			return "true", nil
-		}
-		return "false", nil
+		return constantBool(g, v, t)
 	case compile.ConstantDouble:
-		return fmt.Sprint(float64(v)), nil
+		return constantDouble(g, v, t)
 	case compile.ConstantInt:
-		return fmt.Sprint(int(v)), nil
+		return constantInt(g, v, t)
 	case compile.ConstantList:
-		l, ok := t.(*compile.ListSpec)
-		if !ok {
-			return "", fmt.Errorf(
-				"cannot cast list %v to %v", v, t.ThriftName())
-		}
-		return constantList(g, v, l)
+		return constantList(g, v, t)
 	case compile.ConstantMap:
-		m, ok := t.(*compile.MapSpec)
-		if !ok {
-			return "", fmt.Errorf(
-				"cannot cast map %v to %v", v, t.ThriftName())
-		}
-		return constantMap(g, v, m)
+		return constantMap(g, v, t)
 	case compile.ConstantSet:
-		s, ok := t.(*compile.SetSpec)
-		if !ok {
-			return "", fmt.Errorf(
-				"cannot cast set %v to %v", v, t.ThriftName())
-		}
-		return constantSet(g, v, s)
+		return constantSet(g, v, t)
 	case compile.ConstantString:
 		return strconv.Quote(string(v)), nil
 	case *compile.ConstantStruct:
-		s, ok := t.(*compile.StructSpec)
-		if !ok {
-			return "", fmt.Errorf(
-				"cannot cast struct %v to %v", v, t.ThriftName())
-		}
-		return constantStruct(g, v, s)
+		return constantStruct(g, v, t)
 	case compile.EnumItemReference:
-		return g.TextTemplate(`<typeName .Enum><goCase .Item.Name>`, v)
+		return enumItemReference(g, v, t)
 	case compile.ConstReference:
 		return g.LookupConstantName(v.Target)
 	default:
 		panic(fmt.Sprintf("Unknown constant value %v (%T)", c, c))
 	}
 }
 
-func constantList(g Generator, v compile.ConstantList, l *compile.ListSpec) (string, error) {
+func constantBool(g Generator, v compile.ConstantBool, t compile.TypeSpec) (string, error) {
+	s := "false"
+	if v {
+		s = "true"
+	}
+	if t != compile.BoolSpec {
+		n, err := typeName(g, t)
+		if err != nil {
+			return "", err
+		}
+		s = fmt.Sprintf("%v(%v)", n, s)
+	}
+	return s, nil
+}
+
+func constantDouble(g Generator, v compile.ConstantDouble, t compile.TypeSpec) (string, error) {
+	s := fmt.Sprint(float64(v))
+	if t != compile.DoubleSpec {
+		n, err := typeName(g, t)
+		if err != nil {
+			return "", err
+		}
+		s = fmt.Sprintf("%v(%v)", n, s)
+	}
+	return s, nil
+}
+
+func constantInt(g Generator, v compile.ConstantInt, t compile.TypeSpec) (string, error) {
+	s := fmt.Sprint(int(v))
+	switch t {
+	case compile.I8Spec, compile.I16Spec, compile.I32Spec, compile.I64Spec:
+		// do nothing
+	default:
+		n, err := typeName(g, t)
+		if err != nil {
+			return "", err
+		}
+		s = fmt.Sprintf("%v(%v)", n, s)
```

Comment:
since the casting logic is always the same, should we have a function that does that?
then you could just do,
```go
default:
  return wrapType(g, t, s)
```